### PR TITLE
graph-based segmentation merge

### PR DESCRIPTION
**Pull request by jsundram**
see conversation here: https://discourse.julialang.org/t/hierarchical-merge-on-a-region-adjacency-graph/51633/4

### CODE REVIEW

File: Project.toml
Comment (timholy) [inline Project.toml]: You could avoid a dependence on `Colors` by using `import Images.Colors`. Not an important change, but all dependencies need `[compat]` bounds and the fewer of those we have the easier it is to manage version upgrades.

File: Project.toml
Comment (timholy) [inline Project.toml:13]: This needs `[compat]` bounds added.

File: src/ImageSegmentation.jl
Comment (timholy) [inline src/ImageSegmentation.jl:6]: For the record, this is a 10% increase in `using ImageSegmentation` time. I'm fine with that.

File: src/merge.jl
Comment (timholy) [inline src/merge.jl]: That's very strange. I might investigate after this merges.

File: src/merge.jl
Comment (timholy) [inline src/merge.jl]: Instead of saying `using RegionTrees, MetaGraphs` in ImageSegmentation, say
```
using RegionTrees: RegionTrees, stuff_I_need_from_RegionTrees
using MetaGraphs: MetaGraphs, stuff_I_need_from_MetaGraphs
```

File: src/merge.jl
Comment (timholy) [inline src/merge.jl]:
```suggestion
    function weight(t::Tuple{Edge{Int}, Bool})::Real
```
so it runs properly on 32-bit CPUs. This change should be made everywhere Int64 appears below. (I won't annotate them.)

File: src/merge.jl
Comment (timholy) [inline src/merge.jl]: I'd suggest calling this `invalidate_neighbors!(edge_heap, g, e)` (with a !)

File: src/merge.jl
Comment (timholy) [inline src/merge.jl]: I predict that someday folks will want to generalize this (e.g., pass a weight function), but this is fine for now.

File: test/merge.jl
Comment (timholy) [inline test/merge.jl]: You could leave out the `== true` (`@test stmt` checks whether `stmt` evaluates to `true`), but this is fine as-is.

**Review by timholy — COMMENTED** [review]
Sorry for the delay. This is great! Just minor issues before merging.

**Comment by timholy** [comment]
This is looking good. Two small issues left to discuss:
- it's conventional, but not required, that in mutating functions (ones ending with !), the first argument it typically the mutated entity. There are many exceptions though (e.g., `map!`). Style recommendations are documented here: https://docs.julialang.org/en/v1/manual/style-guide/#Write-functions-with-argument-ordering-similar-to-Julia-Base. Alternatively, you can make a case for the current ordering of `invalidate_neighbors!` and just add an extra note to the docstring about which argument(s) are mutated.
- this exports `merge` which will conflict with `Base.merge`. Since you have unambiguous argument types, should we just extend `Base.merge` rather than creating our own independent `merge` function?

**Comment by jsundram** [comment]
Thanks! re: outstanding points:

1.  Sorry, I didn't notice your original suggestion to change the parameter order for `invalidate_neighbors` in addition to adding a `!` to indicate mutation. Will fix.

2.  It's a good point re: `Base.merge` that I didn't consider when naming the function. The semantics of `Base.merge` function seem to be taking 2 or more instances of a type and returning 1 instance of that type. Which is different than the semantics of `merge` here. The same function is called `merge_hierarchical` [in scikit-image](https://github.com/scikit-image/scikit-image/blob/main/skimage/future/graph/graph_merge.py#L59). I'm not sure if that's better? The intent of this `merge` function is similar to that of `prune_segments` in core.jl -- perhaps `merge_segments` is a reasonable name? I'm happy to follow your preference here, and am totally open to other suggestions. _Would it make sense to rename the file to match whatever function name we settle on?_

**Comment by timholy** [comment]
I appreciate your awareness of the issue "are these two functions really doing the same thing?" That's really important and kudos to you for noticing & caring.

There's one easy solution: don't export `merge`. Then it would be called `ImageSegmentation.merge(args...)`. But you also make a good case for `merge_segments`, I like that quite a lot. I'm fine with whichever you prefer.

The filename is also optional. I'd be happy to see it renamed (that might be clearest), but I'll have no reservations whatsoever if you decide to keep it the same.

File: src/merge_segments.jl
Comment (timholy) [inline src/merge_segments.jl]:
```suggestion
function add_neighboring_edges!(g::MetaGraph, e::AbstractEdge)
    edges = Edge{eltype(e)}[]
```

**Review by timholy — APPROVED** [review]
Sorry, @jsundram, I was about to hit merge and then noticed one more thing. If you're good with this you can just accept the suggestion. Once the tests run, are you OK with merging or is there anything else you want to change?

**Comment by jsundram** [comment]
> Sorry, @jsundram, I was about to hit merge and then noticed one more thing. If you're good with this you can just accept the suggestion. 

Accepted, thanks for that improvement!

> Once the tests run, are you OK with merging or is there anything else you want to change?

I think this looks good. I have a couple of questions that might be better resolved in different PRs:
1. Does it make sense to replace `SimpleWeightedGraph` (in core.jl) with a `MetaGraph` with a property `:weight` on its edges? SimpleWeightedGraph has two issues that I noticed in the [readme](https://github.com/JuliaGraphs/SimpleWeightedGraphs.jl): a) zero-weight edges are discarded by add_edge! and b) adding or removing vertices or edges is not particularly performant. And also this issue https://github.com/JuliaGraphs/SimpleWeightedGraphs.jl/issues/66.
2. Should the list of functions, imports and includes be alphabetized in ImageSegmentation.jl and runtests.jl?

**Comment by timholy** [comment]
Great questions. I'm not very fresh on these things myself, but if memory serves SimpleWeightedGraph is based on sparse matrices and is indeed incredibly slow for additions/removals of edges. If the performance is limiting, I'd be very much in favor of a more performant alternative. But I confess I'm not very fresh on these things so you should take your own experience as a more reliable guide.

Thanks again for a terrific PR, @jsundram!

**Comment by jsundram** [comment]
@timholy -- I was just wondering; do I also need to bump the version number and make a new release, so that people can use the new code easily? I'm not sure how releases get made...

**Comment by timholy** [comment]
My fault, I should have made a release. I guess I might have been holding back wondering if you were wanting to tackle the first item in https://github.com/JuliaImages/ImageSegmentation.jl/pull/64#issuecomment-783616167. It's not required, though; would you like me to make a release?

> I'm not sure how releases get made...

Not much to it:
1. Someone needs to bump the version number. Since this is a new feature it should be 1.5.0 (1.4.9 would be for bugfixes only).
2. Someone with commit privileges needs to notify the bot, see for example https://github.com/JuliaImages/ImageSegmentation.jl/commit/556eabb6e2a075609336b0e4231784914f16455d

How about this: if you're ready for a release now, submit a PR bumping the version; I'll merge it and do the rest. Or tell me to make a new release and I'll do it. (It's no trouble, it's <1min of work total.)

**Comment by jsundram** [comment]
Thanks!

I'll hold off on `SimpleWeightedGraph` -> `MetaGraph` change, since I don't (yet?) have any evidence that there are performance wins to be had as a result of that change. 

I just submitted #65 to get the release train started.

**Pull request merged**
